# Practical Guide to Pandas Essentials for Beginners


This tutorial covers
1. Selecting rows and columns from a dataframe
2. Change values of a dataframe
3. Appending dataframes
4. Handling duplicates
5. Handling missing values
6. Changing columns types

# Quick recap on last tutorial

In last tutorial we have covered basic data operations like read, write, adding new column, recoding existing columns and dropping column. Here is the github link to the jupyter notebook. Also posted below.


# Create the DataFrame

In [100]:
#import libraries
import pandas as pd
from sklearn import datasets

#creating the dataset for this tutorial
iris = datasets.load_iris()
df=pd.DataFrame(iris.data,columns=iris.feature_names)
df['target']=iris['target']

# Use map to recode column
species_mapping={0:iris['target_names'][0],1:iris['target_names'][1],2:iris['target_names'][2]}
#print('Checking species mapping: ')
#print(species_mapping)
#print('\n')

df['Species'] = df['target'].map(species_mapping)
#print("Checking if Species is created : ")
#print(df.head())
#print('\n')

#show few records
print('\n')
print("Iris data shape :", iris['data'].shape)
print('\n')
print("Top records : ")
print(df.head())



Iris data shape : (150, 4)


Top records : 
   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target Species  
0       0  setosa  
1       0  setosa  
2       0  setosa  
3       0  setosa  
4       0  setosa  


# Selecting rows and columns from dataframe

One of the most common data tasks we perform are selecting subsets of data either by selecting set of rows or columns or combination of both. This can be a bit tricky (and frustrating) in pandas unless we know what's happening behind the scenes. Let's see how we can accomplish these tasks in pandas. 
We shall be using the same ```iris``` dataset as used in last turorial. The first part of the script is just recreating the dataset from last tutorial.

## .iloc and .loc

.iloc and .loc are indexers that are used for selecting rows and columns in pandas, .iloc can be used to select rows or columns by number and .loc can used to select rows with row labels or boolean masks. Note that the dataframe created above has row numbers as row labels, this is the default row index when no column is assigned to be the row label of a pandas dataframe. 

When selecting single row or column subsets from a pandas dataframe using double square brackets will return a dataframe and using single square brackets with return a series. Let's see how ```iloc``` can be used to select rows and columns.

# iloc

In [101]:
print("\n")
print("Select row 1 using iloc (returns series) :")
# select 1st row using iloc
df.iloc[0] # returns a series, not specifying colunms returns all columns
df.iloc[0,:] # returns a series, not specifying colunms returns all columns




Select row 1 using iloc (returns series) :


sepal length (cm)       5.1
sepal width (cm)        3.5
petal length (cm)       1.4
petal width (cm)        0.2
target                    0
Species              setosa
Name: 0, dtype: object

In [102]:
print("\n")
print("Select row 1 using iloc (returns dataframe) :")
df.iloc[[0],:] # returns a dataframe, not specifying colunms returns all columns



Select row 1 using iloc (returns dataframe) :


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   

   target Species  
0       0  setosa

In [103]:
print("\n")
print("Select multiple rows using using iloc :")
# select multiple rows using iloc
df.iloc[0:5] # returns a dataframe
df.iloc[0:5,:] # returns a dataframe




Select multiple rows using using iloc :


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target Species  
0       0  setosa  
1       0  setosa  
2       0  setosa  
3       0  setosa  
4       0  setosa

In [104]:
print("\n")
print("Select column 1 using using iloc (returns series) :")
# select 1st column using iloc
df.iloc[:,0] # returns a series, not specifying rows returns all rows
df.iloc[:,0] # returns a series, not specifying rows returns all rows



Select column 1 using using iloc (returns series) :


0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: sepal length (cm), Length: 150, dtype: float64

In [105]:
print("\n")
print("Select column 1 using using iloc (returns dataframe) :")
# select 1st column using iloc
df.iloc[:,[0]] # returns a dataframe, not specifying rows returns all rows



Select column 1 using using iloc (returns dataframe) :


sepal length (cm)
0                  5.1
1                  4.9
2                  4.7
3                  4.6
4                  5.0
..                 ...
145                6.7
146                6.3
147                6.5
148                6.2
149                5.9

[150 rows x 1 columns]

In [106]:
print("\n")
print("Select multiple columns using using iloc :")
# select multiple columns using iloc
df.iloc[:,0:3] # returns a dataframe,not specifying rows returns all rows 



Select multiple columns using using iloc :


sepal length (cm)  sepal width (cm)  petal length (cm)
0                  5.1               3.5                1.4
1                  4.9               3.0                1.4
2                  4.7               3.2                1.3
3                  4.6               3.1                1.5
4                  5.0               3.6                1.4
..                 ...               ...                ...
145                6.7               3.0                5.2
146                6.3               2.5                5.0
147                6.5               3.0                5.2
148                6.2               3.4                5.4
149                5.9               3.0                5.1

[150 rows x 3 columns]

In [107]:
print("\n")
print("Select multiple rows and columns using using iloc :")
# select rows and columns with iloc
df.iloc[0:5,0:3] # returns a dataframe



Select multiple rows and columns using using iloc :


sepal length (cm)  sepal width (cm)  petal length (cm)
0                5.1               3.5                1.4
1                4.9               3.0                1.4
2                4.7               3.2                1.3
3                4.6               3.1                1.5
4                5.0               3.6                1.4

# loc

In [108]:
# select 1st row using loc
print("\n")
print("Select 1 row using loc (returns series) :")
df.loc[0] # returns a series, not specifying colunms returns all columns
df.loc[0,:] # returns a series, not specifying colunms returns all columns



Select 1 row using loc (returns series) :


sepal length (cm)       5.1
sepal width (cm)        3.5
petal length (cm)       1.4
petal width (cm)        0.2
target                    0
Species              setosa
Name: 0, dtype: object

In [109]:
# select 1st row using loc
print("\n")
print("Select 1 row using loc (returns dataframe) :")
df.loc[[0],:] # returns a dataframe, not specifying colunms returns all columns




Select 1 row using loc (returns dataframe) :


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   

   target Species  
0       0  setosa

In [110]:
print("\n")
print("Select multiple rows using loc (returns dataframe) :")
# select multiple rows using loc
df.loc[0:5] # returns a dataframe
df.loc[0:5,:] # returns a dataframe



Select multiple rows using loc (returns dataframe) :


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   
5                5.4               3.9                1.7               0.4   

   target Species  
0       0  setosa  
1       0  setosa  
2       0  setosa  
3       0  setosa  
4       0  setosa  
5       0  setosa

In [111]:
# select 1st column using loc
# df.loc[:,0] # does not work
print("\n")
print("Select 1 column using loc (returns series) :")
df.loc[:,'sepal length (cm)'] # returns a series, not specifying rows returns all rows



Select 1 column using loc (returns series) :


0      5.1
1      4.9
2      4.7
3      4.6
4      5.0
      ... 
145    6.7
146    6.3
147    6.5
148    6.2
149    5.9
Name: sepal length (cm), Length: 150, dtype: float64

In [112]:
# select 1st column using loc
# df.loc[:,0] # does not work
print("\n")
print("Select 1 column using loc (returns dataframe) :")
df.loc[:,['sepal length (cm)']] # returns a dataframe,not specifying rows returns all rows



Select 1 column using loc (returns dataframe) :


sepal length (cm)
0                  5.1
1                  4.9
2                  4.7
3                  4.6
4                  5.0
..                 ...
145                6.7
146                6.3
147                6.5
148                6.2
149                5.9

[150 rows x 1 columns]

In [113]:
# select multiple columns using loc
print("\n")
print("Select multiple columns using loc :")
df.loc[:,['sepal length (cm)','sepal width (cm)']] # returns a dataframe,not specifying rows returns all rows 



Select multiple columns using loc :


sepal length (cm)  sepal width (cm)
0                  5.1               3.5
1                  4.9               3.0
2                  4.7               3.2
3                  4.6               3.1
4                  5.0               3.6
..                 ...               ...
145                6.7               3.0
146                6.3               2.5
147                6.5               3.0
148                6.2               3.4
149                5.9               3.0

[150 rows x 2 columns]

In [114]:
# select rows and columns with loc
print("\n")
print("Select rows and columns using using loc (returns series) :")
df.loc[0:5,'sepal length (cm)'] # returns a series



Select rows and columns using using loc (returns series) :


0    5.1
1    4.9
2    4.7
3    4.6
4    5.0
5    5.4
Name: sepal length (cm), dtype: float64

In [115]:
# select rows and columns with loc
print("\n")
print("Select rows and columns using using loc (returns dataframe) :")
df.loc[0:5,['sepal length (cm)']] # returns a dataframe,not specifying rows returns all rows 




Select rows and columns using using loc (returns dataframe) :


sepal length (cm)
0                5.1
1                4.9
2                4.7
3                4.6
4                5.0
5                5.4

In [116]:
# select rows and columns with loc
print("\n")
print("Select rows and columns using using loc:")
df.loc[0:5,['sepal length (cm)','sepal width (cm)']] # returns a dataframe



Select rows and columns using using loc:


sepal length (cm)  sepal width (cm)
0                5.1               3.5
1                4.9               3.0
2                4.7               3.2
3                4.6               3.1
4                5.0               3.6
5                5.4               3.9

In [117]:
# select rows with conditions using loc
bool_mask = df['sepal length (cm)']>4 # returns series of boolean values
bool_mask

0      True
1      True
2      True
3      True
4      True
       ... 
145    True
146    True
147    True
148    True
149    True
Name: sepal length (cm), Length: 150, dtype: bool

In [118]:
df.loc[bool_mask] #returns dataframe with all columns

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                  5.1               3.5                1.4               0.2   
1                  4.9               3.0                1.4               0.2   
2                  4.7               3.2                1.3               0.2   
3                  4.6               3.1                1.5               0.2   
4                  5.0               3.6                1.4               0.2   
..                 ...               ...                ...               ...   
145                6.7               3.0                5.2               2.3   
146                6.3               2.5                5.0               1.9   
147                6.5               3.0                5.2               2.0   
148                6.2               3.4                5.4               2.3   
149                5.9               3.0                5.1               1.8   

     target    Species  
0         0     setosa  
1         0     setosa  
2         0     setosa  
3         0     setosa  
4         0     setosa  
..      ...        ...  
145       2  virginica  
146       2  virginica  
147       2  virginica  
148       2  virginica  
149       2  virginica  

[150 rows x 6 columns]

In [119]:
df.loc[bool_mask,'target'] #returns series

0      0
1      0
2      0
3      0
4      0
      ..
145    2
146    2
147    2
148    2
149    2
Name: target, Length: 150, dtype: int32

In [120]:
df.loc[bool_mask,['target']] #returns dataframe

target
0         0
1         0
2         0
3         0
4         0
..      ...
145       2
146       2
147       2
148       2
149       2

[150 rows x 1 columns]

In [121]:
# Note the difference
print(".loc")
print(df.loc[0:5,['target']]) # loc returns all rows 0 to 5 with matching labels

print(".iloc")
print(df.iloc[0:5,[4]]) # iloc returns rows 0 to 4, just like regular python indexing in lists, tuples, dictionaries


.loc
   target
0       0
1       0
2       0
3       0
4       0
5       0
.iloc
   target
0       0
1       0
2       0
3       0
4       0


# .Isin()

The ```isin()``` function comes in handy when you need to check for specific values in a row or columns. One way could be to right multiple ```==``` conditions but thet could be really tedious. Let's check an example below.

In [122]:
# select all rows where species is "setosa" or "versicolor"
df_new=df.loc[df['Species'].isin(['setosa','versicolor']),:]
print(df_new.Species.unique())


['setosa' 'versicolor']


In [123]:
# select all rows where species is not "setosa" and "versicolor"
df_new=df.loc[~df['Species'].isin(['setosa','versicolor']),:]
print(df_new.Species.unique())


['virginica']


# Modifying a dataframe
### Views and Copies
Now that we have a good understanding of how to use loc and iloc, lets move on to understanding how to modify a dataframe or create new dataframes by modifying existing dataframes. In order to understand this we need to learn about views and copyies in pandas.Till now we have been creating views of a dataframe which is just a subset of the dataframe, when modyfying a dataframe we need to be careful of wether we are modifying a vew or the original dataframe. Lets check out the example below.

In [124]:
# We want to modify row 5 value for column "target"
# print row 5 and column target
df_copy=df
print("Original value : ",df_copy[4:5][['target']])

# lets modify this value
df_copy[4:5][['target']]=1

# lets check if the value has changed
print("Modified value : ",df_copy.loc[4:4,['target']]) # it did get changed

# print rows where sepal length<=4.5 and column target
print("Original value : ",df_copy[df_copy['sepal length (cm)']<=4.5][['target']])

# lets modify this value
df_copy[df_copy['sepal length (cm)']<=4.5][['target']]=2

# lets check if the value has changed
print("Modified value : ",df_copy[df_copy['sepal length (cm)']<=4.5][['target']]) #it did not get changed


Original value :     target
4       0
Modified value :     target
4       1
Original value :      target
8        0
13       0
38       0
41       0
42       0
Modified value :      target
8        0
13       0
38       0
41       0
42       0


C:\Users\madhura\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys
C:\Users\madhura\AppData\Local\Continuum\anaconda3\lib\site-packages\pandas\core\indexing.py:494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self.obj[item] = s
C:\Users\madhura\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:16: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .lo

When we run this we get ```SettingWithCopyWarning``` as pandas is not sure wether the original dataframe got changed or a view got modified. As we can see the target column in row 4 shows changed value but target column with sepal length <= 4.5 does not change. This warning is a way for pandas to let us know that things might not have gone as expected. We can solve this by using .loc or .iloc and not using chaining.

In [125]:
df_copy=df
print("Original value : ",df_copy.loc[[5],['target']])

# lets modify this value
df_copy.loc[[5],['target']]=2

# lets check if the value has changed
print("Modified value : ",df_copy.loc[[5],['target']]) # it did get changed

# print rows where sepal length<=4.5 and column target
print("Original value : ",df_copy.loc[df_copy['sepal length (cm)']<=4.5,['target']])

# lets modify this value
df_copy.loc[df_copy['sepal length (cm)']<=4.5,['target']]=2

# lets check if the value has changed
print("Modified value : ",df_copy.loc[df_copy['sepal length (cm)']<=4.5,['target']]) # it did get changed


Original value :     target
5       0
Modified value :     target
5       2
Original value :      target
8        0
13       0
38       0
41       0
42       0
Modified value :      target
8        2
13       2
38       2
41       2
42       2


Now lets try to make some changes to subset dataframe using .loc and .iloc.

In [126]:
# create a copy of dataframe
df_copy_sub= df_copy.loc[df_copy['sepal length (cm)']<=4.5,:]

# original value
print("Original value ", df_copy_sub['target'])

# lets modify this value
df_copy_sub['target']=3

# lets check if the value has changed
print("Modified value :",df_copy_sub['target']) 


Original value  8     2
13    2
38    2
41    2
42    2
Name: target, dtype: int32
Modified value : 8     3
13    3
38    3
41    3
42    3
Name: target, dtype: int64


C:\Users\madhura\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


The ```SettingWithCopyWarning``` warning appears again though we have used .loc and .iloc. This is because df_copy is not a copy of df, so when we try to modify df_copy pandas is again confused. To avoid this, better to use ```df_copy=df.copy()``` and not ```df_copy=df```.

In [127]:
df_copy=df.copy()
# save the dataframe
df_copy_sub= df_copy.loc[df_copy['sepal length (cm)']<=4.5,:].copy()

# check original value
print(df_copy_sub['target'])

# lets modify this value
df_copy_sub['target']=3

# lets check if the value has changed
print(df_copy_sub['target']) # it get changed

8     2
13    2
38    2
41    2
42    2
Name: target, dtype: int32
8     3
13    3
38    3
41    3
42    3
Name: target, dtype: int64


Finally, we got rid of the pesky ```SettingWithCopyWarning```.

# Appending dataframes

It is common to append 2 dataframes on columns or rows while working our way through data and pandas provides a simple way to do so with the ```concat``` function.

### Dataframes can be appended on column by using option ``` axis=1``` 

In [128]:
# Lets create a copy of existing data frame "df"
df_1= df.copy()
df_2= df.copy()

# Lets append by column
df_append_bycol= pd.concat([df_1,df_2],axis=1)
#df_append_bycol= pd.concat([df_1,df_2],axis="columns") #just another way


print("Original Dataframe - ",df.shape) # original dataframe has 150 records and 5 columns
print("Appended Dataframe - ",df_append_bycol.shape) # new dataframe has 150 records and 10 columns

print("Concatenated Dataframe")
print(df_append_bycol.head(20)) # new dataframe has 150 records and 10 columns

Original Dataframe -  (150, 6)
Appended Dataframe -  (150, 12)
Concatenated Dataframe
    sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4             

### Unequal rows
Check what happens when 2 dataframes have unequal number of rows. In the example below first dataframe has only 6 rows while second dataframe has 150 rows. The concatenated dataframe has NaN values in the first 5 columns (coming from first dataframe) in all rows beyound row 6.

In [129]:
df_1= df.loc[0:5,:].copy() # df_1 has only 6 rows
df_2= df.copy()

# Lets append by column
df_append_bycol= pd.concat([df_1,df_2],axis=1)

print("Original Dataframe - ",df.shape) # original dataframe has 150 records and 5 columns
print("Appended Dataframe - ",df_append_bycol.shape) # new dataframe has 150 records and 10 columns
print("\n")

# Notice the first 5 columns are NaN from row 7 onwards
print("Concatenated Dataframe")
print(df_append_bycol.head(10)) 

Original Dataframe -  (150, 6)
Appended Dataframe -  (150, 12)


Concatenated Dataframe
   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   
5                5.4               3.9                1.7               0.4   
6                NaN               NaN                NaN               NaN   
7                NaN               NaN                NaN               NaN   
8                NaN               NaN                NaN               NaN   
9                NaN               NaN                NaN               NaN   

   target Species  sepal length (cm)  sepa

### Different row index
Check what happens when 2 dataframes have different row indexes. In the example below first dataframe has row index 5 to 10 while second dataframe has row index 0 to 149. The concatenated dataframe has NaN values in the first 5 columns (coming from first dataframe) in all rows but rows 5 to 10. ```concat``` function merged the 2 dataframes on row index, so when concatenating dataframes we need to keep any eye on the row index as well. This can be solved by reseting index first for dataframe.

In [130]:
df_1= df.loc[5:10,:].copy() # df_1 has only 6 to 10 from original dataframe
df_2= df.copy()

# Lets append by column
df_append_bycol= pd.concat([df_1,df_2],axis=1)

print("Original Dataframe - ",df.shape)
print("Appended Dataframe - ",df_append_bycol.shape)
print("\n")

# Notice the first 5 columns are NaN from row 7 onwards
print("Concatenated Dataframe")
print(df_append_bycol.head(20)) 
print("\n")

# reset index
df_1= df.loc[5:10,:].copy().reset_index(drop=True) # df_1 has only 6 to 10 from original dataframe
df_2= df.copy()

# Lets append by column
df_append_bycol= pd.concat([df_1,df_2],axis=1)

print("Original Dataframe - ",df.shape)
print("Appended Dataframe - ",df_append_bycol.shape)
print("\n")

# Notice the first 5 columns are NaN from row 7 onwards
print("Concatenated Dataframe with reset index")
print(df_append_bycol.head(20)) 

Original Dataframe -  (150, 6)
Appended Dataframe -  (150, 12)


Concatenated Dataframe
    sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 NaN               NaN                NaN               NaN   
1                 NaN               NaN                NaN               NaN   
2                 NaN               NaN                NaN               NaN   
3                 NaN               NaN                NaN               NaN   
4                 NaN               NaN                NaN               NaN   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4           

### Dataframes can be appended on row by using option ``` axis=0``` 
When appending dataframes by row, the row indexes also get appended and needs to be reset.

In [131]:
# Lets create a copy of existing data frame "df"
df_1= df.loc[0:5,:].copy()
df_2= df.loc[0:5,:].copy()

# Lets append by row
df_append_byrow= pd.concat([df_1,df_2],axis=0) # check the row index
print("\n")
print("Original Dataframe - ",df_1.shape)
print("Appended Dataframe - ",df_append_byrow.shape)
print("\n")
print("Concatenated Dataframe without reset index")
print(df_append_byrow.head(10))

# Lets append by row and reset index
df_append_byrow= pd.concat([df_1,df_2],axis=0).reset_index(drop=True) # check the row index
print("\n")
print("Original Dataframe - ",df_1.shape)
print("Appended Dataframe - ",df_append_byrow.shape)
print("\n")

# check the row index
print("Concatenated Dataframe with reset index")
print(df_append_byrow.head(10))



Original Dataframe -  (6, 6)
Appended Dataframe -  (12, 6)


Concatenated Dataframe without reset index
   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   
5                5.4               3.9                1.7               0.4   
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   

   target Species  
0   

### Unequal columns
Check what happens when 2 dataframes have unequal number of columns. In the example below first dataframe has only 3 columns while second dataframe has 5 columns. The concatenated dataframe has NaN values in the first 10 rows (coming from first dataframe) in columns that are missing from it.

Interesting fact !
Notice that in the concatenated dataframe the order of columns has changed, ```concat``` reorders the columns in alphabetical order, use ```sort= False``` to turn this off.

In [132]:
# lets take a smaller subset to be able to inspect the results
df_1= df.iloc[0:10,0:1].copy() # df_1 has only 6 rows
df_2= df.iloc[0:10,:].copy()

print(df_1.head())
print(df_2.head())

# Lets append by row
df_append_byrow= pd.concat([df_1,df_2],axis=0,sort=False)

print("\n")
print("Original Dataframe1 - ",df_1.shape)
print("Original Dataframe2 - ",df_2.shape)

print("Appended Dataframe - ",df_append_byrow.shape)
print("\n")

# Notice the first 5 columns are NaN from row 7 onwards
print("Concatenated Dataframe")
print(df_append_byrow.head(20)) # new dataframe has 150 records and 10 columns

   sepal length (cm)
0                5.1
1                4.9
2                4.7
3                4.6
4                5.0
   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target Species  
0       0  setosa  
1       0  setosa  
2       0  setosa  
3       0  setosa  
4       1  setosa  


Original Dataframe1 -  (10, 1)
Original Dataframe2 -  (10, 6)
Appended Dataframe -  (20, 6)


Concatenated Dataframe
   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               NaN                NaN               NaN   
1 

## Handling duplicates
It's one of basic data sanity checks to look for duplicate rows and columns and treat them appropriately before commencing any analysis on data. And you guessed it right, pandas has a nifty function to do this - ```drop_duplicates()```. With this function we can either **drop all duplicates** or **keep the first/last duplicate** (after ordering by some other columns). We can also consider only a subset of columns when searching for duplicates by using ``` subset ``` option. Lets see how this works.


In [133]:
# lets first add some duplicate rows to our dataframe
df_1= df.loc[0:5,:].copy()
df_2= df.loc[0:5,:].copy()
df_hasduplicates= pd.concat([df_1,df_2],axis=0).reset_index(drop=True) # this dataframe has duplicate rows

In [134]:
print("Dataframe with duplicates ",df_hasduplicates.shape)
#print(df_hasduplicates.head(5))

# drop them all
df_noduplicates= df_hasduplicates.drop_duplicates()
print("Dataframe without duplicates ",df_noduplicates.shape)
#print(df_noduplicates.head(5))


Dataframe with duplicates  (12, 6)
Dataframe without duplicates  (6, 6)


In [135]:
# let's check how many duplicate values are there in target column
# we can check using group by function (will be covered in a later article)
print("Dataframe with duplicates")
print(df_hasduplicates['target'].value_counts())
print("\n")

# drop the rows with duplicate target column
df_noduplicates_1= df_hasduplicates.drop_duplicates(subset=['target'])
#print("Dataframe without duplicates ")
# print(df_noduplicates.shape) #default is to keep the first row
# print("\n")

# let's check if target column has duplicates now
print("Dataframe without duplicates ")
print(df_noduplicates_1['target'].value_counts())
print("\n")

Dataframe with duplicates
0    8
2    2
1    2
Name: target, dtype: int64


Dataframe without duplicates 
2    1
1    1
0    1
Name: target, dtype: int64




We can see that the dataframe with duplicate values has 8 records with value 0, 2 records with value 1 and 2 records with value 2 and the dataframe without duplicates has 1 value of each type. Note, that here pandas keeps the first record for each value of target column.
Next, lets try to use the option ```keep``` to keep records with lowest sepal length.

In [136]:
# drop the rows with duplicate target column while keeping the row with highest sepal length
# first sort by target and petal length
df_hasduplicates.sort_values(['target','sepal length (cm)'],ascending=False,inplace=True) 

#drop duplicates
df_noduplicates_2= df_hasduplicates.drop_duplicates(subset=['target'],keep='last')
#print("Dataframe without duplicates ",df_noduplicates.shape)

# let's check if target column has duplicates now
print("Dataframe without duplicates ")
print(df_noduplicates_2['target'].value_counts())
print("\n")


Dataframe without duplicates 
2    1
1    1
0    1
Name: target, dtype: int64




In [137]:
print("First dataframe :")
print(df_noduplicates_1)
print("\n")

print("Second dataframe :")
print(df_noduplicates_2)
print("\n")


First dataframe :
   sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
4                5.0               3.6                1.4               0.2   
5                5.4               3.9                1.7               0.4   

   target Species  
0       0  setosa  
4       1  setosa  
5       2  setosa  


Second dataframe :
    sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
11                5.4               3.9                1.7               0.4   
10                5.0               3.6                1.4               0.2   
9                 4.6               3.1                1.5               0.2   

    target Species  
11       2  setosa  
10       1  setosa  
9        0  setosa  




# Handling missing values

Knowing how to find and replace missing values is key to data analysis. In pandas we can filter out both rows and columns with any or all missing values with the ```dropna()``` function. 
Let's see the different ways in which pandas lets us manage missing values. 

In [138]:
import numpy as np

# lets first introduce some missing values into our dataframe
df_copy = df.copy()
print(df_copy.shape)
# lets randomly pick some rows and column positions and assign them to NaN
np.random.seed(200)
row=np.random.randint(0,df_copy.shape[0]-1,10)
col=np.random.randint(0,df_copy.shape[1]-1,1)
print(row,col)
df_copy.iloc[row,col]=np.nan


(150, 6)
[ 26 105  16  68  42  76  79  89  14  89] [3]


In [139]:
# Check how many missing values in total
print("No. of missing values :",df_copy.isnull().sum().sum())
print("\n")
# Check how many missing values by columns
print("Missing values summary :")
print(df_copy.isnull().sum())

No. of missing values : 9


Missing values summary :
sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     9
target               0
Species              0
dtype: int64


In [140]:
# Drop rows with any missing values 
df_nonull=df_copy.dropna(how='any')
print(df_nonull.isnull().sum())

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     0
target               0
Species              0
dtype: int64


In [141]:
# Drop rows with missing values in sepal length column
df_nonull=df_copy.dropna(subset=['sepal length (cm)'],how='any')
print(df_nonull.isnull().sum()) 

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     9
target               0
Species              0
dtype: int64


In [142]:
# get all columns with no null values
df_nonull=df_copy.loc[:,df_copy.notnull().all()]
print(df_nonull.isnull().sum()) 
print("Columns with all null values:", df_nonull.columns.values)

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
target               0
Species              0
dtype: int64
Columns with all null values: ['sepal length (cm)' 'sepal width (cm)' 'petal length (cm)' 'target'
 'Species']


In [143]:
# get all columns with any non-null values
df_nonull=df_copy.loc[:,df_copy.notnull().any()]
print(df_nonull.isnull().sum()) 
print("Columns with any non-null values:", df_nonull.columns.values)

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     9
target               0
Species              0
dtype: int64
Columns with any non-null values: ['sepal length (cm)' 'sepal width (cm)' 'petal length (cm)'
 'petal width (cm)' 'target' 'Species']


In [144]:
# get all columns with all null values
df_nonull=df_copy.loc[:,df_copy.isnull().all()]
print(df_nonull.isnull().sum()) 
print("Columns with all null values:", df_nonull.columns.values)

Series([], dtype: float64)
Columns with all null values: []


In [145]:
# get all columns with all null values
df_nonull=df_copy.loc[:,df_copy.isnull().any()]
print(df_nonull.isnull().sum()) 
print("Columns with any null values:", df_nonull.columns.values)


petal width (cm)    9
dtype: int64
Columns with any null values: ['petal width (cm)']


In [146]:
# Get all columns with all non-zero values
df_nonull=df_copy.loc[:,df_copy.all()]
print(df_nonull.isnull().sum()) 
print("Columns with all non-zero values:", df_nonull.columns.values)

# get all columns with any non-zero values
df_nonull=df_copy.loc[:,df_copy.any()]
print(df_nonull.isnull().sum()) 
print("Columns with any non-zero values:", df_nonull.columns.values)

sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     9
Species              0
dtype: int64
Columns with all non-zero values: ['sepal length (cm)' 'sepal width (cm)' 'petal length (cm)'
 'petal width (cm)' 'Species']
sepal length (cm)    0
sepal width (cm)     0
petal length (cm)    0
petal width (cm)     9
target               0
Species              0
dtype: int64
Columns with any non-zero values: ['sepal length (cm)' 'sepal width (cm)' 'petal length (cm)'
 'petal width (cm)' 'target' 'Species']


# Changing datatypes
In pandas the most commonly used datatypes are integer, float, object, category and datetime. We can get the data types of all columns in any dataframe by using the ```.dtypes``` fucntion. It's quite common for data to get read in format not appropriate for analysis (especially when reading from sql database) and needs to be changed to appropriate data type before analysis.For converting to numeric we can use ```astype()``` function or ```to_numeric()``` function which comes in handy especially when you have missing values. There are situations where we need to just select columns of a certain datatype to perform analysis, pandas has a very useful function ``` select_dtypes()``` to do so as well. Let's learn with examples. 


In [147]:
# check datatypes in our dataframe
print(df.dtypes)

sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
target                 int32
Species               object
dtype: object


In [148]:
# lets select all float columns
df_float = df.select_dtypes(include='float64')
print(df_float.dtypes)

sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
dtype: object


In [149]:
# lets convert target column to float type
df['target']=df['target'].astype('float64')
print(df.dtypes)

sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
target               float64
Species               object
dtype: object


In [150]:
# lets convert target column  back to int type
df['target']=df['target'].astype('int64')
print(df.dtypes)

sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
target                 int64
Species               object
dtype: object


In [151]:
# Lets introduce some missing values in target column
row=np.random.randint(0,df.shape[0],5)
df_copy=df.copy()
df_copy.loc[row,'target']=np.nan

#check for missing values
print("Missing values :",df_copy['target'].isnull().sum())


Missing values : 5


In [152]:
# lets convert target column to int type using astype()
# You get an Error !!
df_copy['target']=df_copy['target'].astype(int)

ValueError: Cannot convert non-finite values (NA or inf) to integer

In [153]:
# Now try this
df_copy['target']=pd.to_numeric(df_copy['target'],errors='coerce')
print(df_copy.dtypes)

# or this
df_copy['target']=df_copy['target'].astype(float)

sepal length (cm)    float64
sepal width (cm)     float64
petal length (cm)    float64
petal width (cm)     float64
target               float64
Species               object
dtype: object


That's all for this tutorial. I hope you find it useful and thanks for reading !